presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -68,6 +72,24 @@ public HivePageSourceProvider(
     @Override
     public ConnectorPageSource createPageSource(ConnectorSession session, ConnectorSplit split, List<ColumnHandle> columns)
     {
+        if (HiveSessionProperties.getReadAsQueryUser(session)) {
+            UserGroupInformation ugi = UgiUtils.getUgi(session.getUser());
+            try {
+                return ugi.doAs((PrivilegedExceptionAction<ConnectorPageSource>) () ->
```

Comment:
@dain if the a single query accesses this method repeatedly with multiple threads we should be ok since the query will always be executed by the same user and the method will always doAs that user. We'd have problems if the ConnectorPageSource returned by this method was then shared across queries/users. Would that ever be the case?

In an initial version of this patch we ran into user/threading issues in the BackgroundHiveSplitLoader because we were doing doAs at too high a level before kicking off tasks in the threadpool. This had the effect of binding the subject to the thread across queries. The fix was to do doAs within the thread, not outside of it.

We've been successfully running this patch in production for a week or so without issue, but that doesn't mean there couldn't be a corner case that we've missed.